### ISSUE TITLE
Gradient computation with backpropagation slower than param-shift with TensorFlow interface

### ISSUE THREAD

**Issue by mariaschuld** [PennyLaneAI/pennylane#1244]
We stumbled across an unexpected TensorFlow slowdown when computing a gradient with a backpropagation qnode.

For example, the problem changed the results of [this demo](https://pennylane.ai/qml/demos/tutorial_backprop.html), which is supposed to show that backpropagation speeds up the calculation of gradients. But somehow it does not do so any more when using TensorFlow. Rewriting the demo using autograd leads to expected behaviour.  

This could be related to #1176, however if I read correctly, the issue there speaks about a forward pass, not the comparison of two differentiation rules in the same interface. 

This is a small example:

```python

import pennylane as qml
import timeit
import tensorflow as tf

dev = qml.device("default.qubit", wires=4)

def circuit(params):
    qml.RX(params[0], wires=0)
    return qml.expval(qml.PauliZ(0))

reps = 200
num = 3
params = tf.Variable([0.1])

# parameter-shift
qnode_shift = qml.QNode(circuit, dev, interface="tf", diff_method="parameter-shift")
with tf.GradientTape(persistent=True) as tape:
    res = qnode_shift(params)
t = timeit.repeat("tape.gradient(res, params)", globals=globals(), number=num, repeat=reps)
print("param-shift", min(t) / num)

# backprop
qnode_backprop = qml.QNode(circuit, dev, interface="tf", diff_method="backprop")
with tf.GradientTape(persistent=True) as tape:
    res = qnode_backprop(params)
t = timeit.repeat("tape.gradient(res, params)", globals=globals(), number=num, repeat=reps)
print("backprop", min(t) / num)
```
which produces 

```
ps 0.00045036566674146644
bp 0.0020053336666023824
```
on my PC.

**Comment by josh146** [comment]
Thanks @mariaschuld! Will look into this more when I get a chance. Out of curiosity, was this change picked up by ASV?

**Comment by mariaschuld** [comment]
I haven't looked systematically at results in a while, but when I ran the benchmarks on branches around the introduction of the switch to backpropagation I was super puzzled that I couldn't reproduce the speedup. I pinged @albi3ro and put it onto my todo list, but haven't had time to tackle it yet. 

Maybe there is something really deep going on here, if all of these issues are connected?

**Comment by albi3ro** [comment]
For the data below, I used a different circuit with more variables:

```python
import pennylane as qml
import tensorflow as tf
from pennylane import numpy as np

dev = qml.device("default.qubit", wires=4)

def circuit(params):
    qml.templates.StronglyEntanglingLayers(params, wires=[0,1,2,3])
    return qml.expval(qml.PauliZ(0) @ qml.PauliZ(1) @ qml.PauliZ(2) @ qml.PauliZ(3))

reps = 200
num = 3
rng = np.random.default_rng(seed=42)
params = rng.standard_normal(qml.templates.StronglyEntanglingLayers.shape(2,4))
params = tf.Variable(params)
```

but otherwise performed similar timing.

You just used `min(t)` in your example.  You get very different information if you instead use `max`:

```
param-shift min:  0.0029535000000275127
param-shift max:  0.04009366666650749
backprop min:  0.024844700000054825
backprop max:  0.04199909999988449
```

We get a lot more information when we plot the `t` variables coming out from `timeit`: 

![image](https://user-images.githubusercontent.com/6364575/116123966-47c8bb80-a689-11eb-966f-3c5db038d050.png)

The first time the gradient is computed from parameter-shift, the execution takes a order of magnitude longer and comparable time to backprop.  Then the gradient computation continues at every low number.

This indicates parameter-shift is caching something.

**Comment by josh146** [comment]
Thanks @albi3ro! Great detective work, that really helps pin down the issue.

I understand now what is happening. This is related to the change introduced in #1110, whereby the Jacobian and Hessian matrices are cached within the `tf.custom_gradient` decorator. This was done due to an inefficiency in how PennyLane integrates with the machine learning frameworks: rather than returning the VJP or VHP, PennyLane computes the **full Hessian and Jacobian** with every backwards pass, which is potentially very wasteful.

As a fix, the Jacobian and Hessian are cached on the backward pass. Notably, the cache is (intentionally) cleared on the forward pass, so this caching is very short-term, and only intended to have an effect on consecutive backward pass calls.

In the corresponding QML demo, the demo was written such that:

- The forward pass is computed *once*, and then
- The backward pass is repeated multiple times and timed.

Since the forward pass is not re-computed with different parameter values, the full Jacobian matrix is not recomputed! 

A couple of notes:

1. I would say this is expected behaviour; the caching is doing what it should be doing. If we were to modify the demo so that the backward pass loop is performed with different parameters each time, we will observe the original (and expected) timings.

2. Why don't we see the same behaviour with Autograd? This is because of Autograd's pure functional interface. `qml.grad(qnode)(params)` is a pure function, and does not save the state; it always implicitly includes a forward pass before the backward pass is computed to reconstruct the computational graph. So when we are timing `qml.grad(qnode)(params)`, the cache is being reset between every call.

  Compare this to TensorFlow, where object-oriented gradient UI explicitly separates the forward pass from the backward pass, allowing the backward pass to be executed numerous times for the same computational graph.

So I would say this issue can be closed -- everything is working as expected, and the differences in behaviour between TensorFlow and Autograd come down to the benchmarks itself: what we are timing in autograd (the forward + backward pass) is not comparable to what we are timing in TensorFlow (just the backward pass for the same computational graph).

**Comment by mariaschuld** [comment]
Nice, happy this is solved. We will have to make sure that our benchmarks elsewhere take this into account!

**Comment by josh146** [comment]
Yep. I think the takeaway is that we need to be more careful with how we benchmark interfaces. In this particular example, benchmarking just `qml.grad()` vs. TensorFlow's `tape.gradient()` is not equivalent, since the former involves more computation.

Previously, this was not obvious because the backward pass always drowned out the forward pass, but the caching makes it more obvious!
